Fixes inconsistent inlet and outlet air flow rates for FanCoil, Series PIU, EvapCooler, and HPWH

rraustad opened the pull request:
## Pull request overview

Review of existing example files for inconsistent air flow at inlet and outlet of HVAC equipment shows that at times, the outlet and inlet node mass flow rates do not match. This effort is a result of investigating https://github.com/NREL/EnergyPlus/issues/5768. For the HPWH the unit mass flow rate was calculated using the density of air at the rating conditions and then using actual density during the simulation which led to slight differences in unit inlet and outlet air flow. For evap coolers, the fan was allowed to have a max air volume flow rate less than the evap cooler which led to outlet air flow lower than inlet air flow. For series PIU and fan coil units, the models did not account for the fan availability schedule when selecting operating air flow rate which led to outlet node air flow of 0 while the inlet air flow was set at the operating air flow.

### Work Checklist

Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
- [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
- [x] At least one of the following appropriate labels must be added to this PR to be consumed into the changelog:
  - Defect: This pull request repairs a github defect issue.  The github issue should be referenced in the PR description
### Review Checklist

This will not be exhaustively relevant to every PR.
- [ ] Code style (parentheses padding, variable names)
- [ ] Functional code review (it has to work!)
- [x] If defect, results of running current develop vs this branch should exhibit the fix
- [x] CI status: all green or justified
- [ ] Performance: CI Linux results include performance check -- verify this
- [x] Unit Test(s)
- C++ checks:
  - [ ] Argument types
  - [ ] If any virtual classes, ensure virtual destructor included, other things
- IDD changes:
  - [ ] Verify naming conventions and styles, memos and notes and defaults
  - [ ] Open windows IDF Editor with modified IDD to check for errors
  - [ ] If transition, add rules to spreadsheet
  - [ ] If transition, add transition source
  - [ ] If transition, update idfs
- [ ] If new idf included, locally check the err file and other outputs
- [ ] Documentation changes in place
- [ ] ExpandObjects changes??
- [ ] If output changes, including tabular output structure, add to output rules file for interfaces

rraustad committed:
Fixes for FanCoil, EvapCooler, and HPWH (5 of 10)

rraustad committed:
Move mass conservation warnings

rraustad commented:
@wfbuhl what is this code trying to calculate? If the fan is off, the flow is 0 and the resulting Q should also be 0 ?? Is this code trying to calculate heating output with flow but without fan heat?

line 3287 of SingleDuct.cc

```
// region 3: active heating with fan off
FanOp = 0;
CalcVAVVS( SysNum, FirstHVACIteration, ZoneNodeNum, HCType, MaxFlowWater, BigLoad, FanType, MinMassFlow, FanOp, QHeatFanOffMax );
CalcVAVVS( SysNum, FirstHVACIteration, ZoneNodeNum, HCType, MinFlowWater, 0.0, FanType, MinMassFlow, FanOp, QNoHeatFanOff );
```

rraustad committed:
Show warnings only once per timestep

wfbuhl commented:
Unh, with the fan off there is still air flow because the plenum is
slightly pressurized?

Fred

On Tue, Jul 12, 2016 at 12:54 PM, Richard Raustad notifications@github.com
wrote:

> @wfbuhl https://github.com/wfbuhl what is this code trying to
> calculate? If the fan is off, the flow is 0 and the resulting Q should also
> be 0 ?? Is this code trying to calculate heating output with flow but
> without fan heat?
> 
> // region 3: active heating with fan off
> FanOp = 0;
> CalcVAVVS( SysNum, FirstHVACIteration, ZoneNodeNum, HCType, MaxFlowWater, BigLoad, FanType, MinMassFlow, FanOp, QHeatFanOffMax );
> CalcVAVVS( SysNum, FirstHVACIteration, ZoneNodeNum, HCType, MinFlowWater, 0.0, FanType, MinMassFlow, FanOp, QNoHeatFanOff );
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> https://github.com/NREL/EnergyPlus/pull/5771#issuecomment-232160117, or mute
> the thread
> https://github.com/notifications/unsubscribe/AGHXJDrtmiDhlR7E_6_5QwdeuZQUNkHlks5qU_EBgaJpZM4JJxR4
> .

rraustad commented:
Not from what I see at the end of each fan type calc routine:

```
} else {
    //Fan is off and not operating no power consumed and mass flow rate.
    Fan( FanNum ).FanPower = 0.0;
    FanShaftPower = 0.0;
    PowerLossToAir = 0.0;
    Fan( FanNum ).OutletAirMassFlowRate = 0.0;
```

rraustad commented:
For HVACTemplate-5ZoneVAVFanPowered, the fan schedule is more restrictive than the TU availability schedule. The PIU sets flow if the TU is available, then the fan shuts down the flow. For this reason the node inlet and outlet flow rates do not match. The question is, if the fan is scheduled off, can there possibly be flow through a PIU? My guess is no. I can fix this (and will) but would like confirmation that both the PIU avail sched and the fan sched should be tested before flow is set on the PIU inlet node.

AirTerminal:SingleDuct:SeriesPIU:Reheat,
  SPACE1-1 VAV Series PIU Reheat,                          !- Name
  HVACTemplate-Always 1,                                   !- Availability Schedule Name

Fan:ConstantVolume,
  SPACE1-1 PIU Fan,                                        !- Name
  FanAvailSched,                                           !- Availability Schedule Name

![piufanflow](https://cloud.githubusercontent.com/assets/5778156/16812539/9075a910-48fb-11e6-9c54-2bf7e6145be7.png)

Including a test for fan availability schedule, only for the series PIU, yields proper node flow with no node outlet/inlet mismatch flow warnings:

![piufanflow_fixed](https://cloud.githubusercontent.com/assets/5778156/16813200/7cd1091a-48fe-11e6-8bf6-a9c9bf103f13.png)

mjwitte commented:
Isn't the airloop fan also off at that time, so there shouldn't be any flow at all. If the terminal unit schedule is off, then yes the flow should be zero, but if the terminal unit schedule is on, that doesn't really mean anything.  Then it's up to the airloop fan and the PIU fan whether there's flow.  And for the piu fan, you can't just check the schedule, because it may be overridden by an availability manager.

rraustad committed:
Added check for fan avail schedule for PIU

EnergyArchmage commented on src/EnergyPlus/HVACManager.cc:
```diff
@@ -2725,6 +2732,182 @@ namespace HVACManager {
+
+		// SUBROUTINE LOCAL VARIABLE DECLARATIONS:
+		static Real64 CurrentEndTime( 0.0 ); // end time of time step for current simulation time step
+
```
We don't want new static variables buried inside routines.  This needs to be moved up so it can be reinitialized in clear_state()

rraustad committed:
Move static to clear_state

rraustad committed:
Correct AirTerminalSingleDuctSeriesPIUReheat_GetInputtest unit test

rraustad committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into #5768-Example-files-which-show-discontinutity-in-air-or-water-flow-rate-through-fans-and-pumps

rraustad commented:
Seeing diff's in FanCoil_HybridVent_VentSch. There are warmup diff's that cause a slight difference in results. Pretty much the same operation, just a slight difference in when the vent availability manager turns on.

![5768_fancoilavailmanager](https://cloud.githubusercontent.com/assets/5778156/16999337/78970772-4e8b-11e6-9d51-df573388ee09.png)

nrel-bot-3 commented:
@rraustad @lgentile it has been 16 days since this pull request was last updated.

Myoldmopar commented:
@rraustad Could you update the PR description here?  At least the first little section.  The PR is linked to from the ChangeLog so users will come to that description to see what changes were made.

rraustad commented:
I need to move the mass balance checks out of this branch, or move the fixes to a new branch, before this can be merged.

rraustad committed:
Remove fan and pump node mass flow checks

rraustad commented on src/EnergyPlus/HVACManager.cc:
```diff
@@ -2373,18 +2372,30 @@ namespace HVACManager {
+					MixSenLoad( ZoneLoop ) += CrossMixing( MixNum ).DesiredAirFlowRate * AirDensity * CpAir * ( MAT( ZoneLoop ) - MAT( CrossMixing( MixNum ).ZonePtr ) );
+					H2OHtOfVap = PsyHgAirFnWTdb( ( ZoneAirHumRat( ZoneLoop ) + ZoneAirHumRat( CrossMixing( MixNum ).ZonePtr ) ) / 2.0, ( MAT( ZoneLoop ) + MAT( CrossMixing( MixNum ).ZonePtr ) ) / 2.0 );
+					MixLatLoad( ZoneLoop ) += CrossMixing( MixNum ).DesiredAirFlowRate * AirDensity * ( ZoneAirHumRat( ZoneLoop ) - ZoneAirHumRat( CrossMixing( MixNum ).ZonePtr ) ) * H2OHtOfVap;
+				}
```
I must have made a mistake, there should be no diff's in HVACManager.

rraustad replied:
Well this is what is in develop so these changes are accurate, just not sure why they show as diff's in this branch.

Myoldmopar replied:
What...the... ❓ 

This is weird.

rraustad committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into #5768-Example-files-which-show-discontinutity-in-air-or-water-flow-rate-through-fans-and-pumps

rraustad commented:
I don't know what to do to fix this. My local branch says I'm up to date with the remote branch and with develop? The latest results above show an obvious problem.

Myoldmopar commented:
I'll check it again after lunch. Real confused...

mjwitte commented:
Is it possibly a line-ending mis-match?

Myoldmopar commented:
That was my initial concern, but I don't think so.  I mean, it is showing that those lines _aren't_ in develop at all in this diff.

rraustad commented:
What I did was switch to develop branch and pull. Then saved a few files to overwrite this branch (Fans.hh, Pumps.hh, HVACManager.hh and HVACManager.cc). Then switched to this branch, pulled develop, and then copy those 4 files here. Then pushed those changes. It all worked except for the diff's in HVACManager. And the fact that this branch now shows 256 behind develop which doesn't make sense. In any event, I don't know how to fix it.

Myoldmopar committed:
Merge remote-tracking branch 'origin/develop' into #5768-Example-files-which-show-discontinutity-in-air-or-water-flow-rate-through-fans-and-pumps

Myoldmopar commented:
Merge problem fixed.  Am I to review this?

rraustad commented:
You are welcome to review and provide comments.
Does the CI need to run again? i.e., are the previous CI results valid?

Myoldmopar commented:
Well, it'll run anyway at this point, so no reason to merge it until it is done.

rraustad committed:
Merge branch '#5768-Example-files-which-show-discontinutity-in-air-or-water-flow-rate-through-fans-and-pumps' of https://github.com/NREL/EnergyPlus into #5768-Example-files-which-show-discontinutity-in-air-or-water-flow-rate-through-fans-and-pumps

rraustad committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into #5768-Example-files-which-show-discontinutity-in-air-or-water-flow-rate-through-fans-and-pumps

Conflicts:
	src/EnergyPlus/WaterThermalTanks.hh

nrel-bot-3 commented:
@rraustad @lgentile it has been 14 days since this pull request was last updated.

nrel-bot-3 commented:
@rraustad @lgentile it has been 14 days since this pull request was last updated.

nrel-bot-3 commented:
@rraustad @lgentile it has been 14 days since this pull request was last updated.

nrel-bot-3 commented:
@rraustad @lgentile it has been 14 days since this pull request was last updated.

nrel-bot-3 commented:
@rraustad @lgentile it has been 14 days since this pull request was last updated.

nrel-bot-3 commented:
@rraustad @lgentile it has been 14 days since this pull request was last updated.

nrel-bot-3 commented:
@rraustad @lgentile it has been 14 days since this pull request was last updated.

nrel-bot-3 commented:
@rraustad @lgentile it has been 14 days since this pull request was last updated.

nrel-bot-3 commented:
@rraustad @lgentile it has been 21 days since this pull request was last updated.

nrel-bot-3 commented:
@rraustad @lgentile it has been 15 days since this pull request was last updated.

nrel-bot-3 commented:
@rraustad @lgentile it has been 14 days since this pull request was last updated.

nrel-bot-3 commented:
@rraustad @lgentile it has been 14 days since this pull request was last updated.

nrel-bot-3 commented:
@rraustad @lgentile it has been 14 days since this pull request was last updated.

rraustad committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into #5768-Example-files-which-show-discontinutity-in-air-or-water-flow-rate-through-fans-and-pumps

# Conflicts:
#	src/EnergyPlus/PoweredInductionUnits.cc
#	testfiles/StripMallZoneEvapCoolerAutosized.idf
#	tst/EnergyPlus/unit/AirTerminalSingleDuctPIUReheat.unit.cc

rraustad commented:
This one slipped by. Still have issues with inlet/outlet flow mismatch in fan and pumps for some example files.

![5768-fancoil_hybridvent_ventsch](https://cloud.githubusercontent.com/assets/5778156/24508992/23152d46-1533-11e7-9e46-366bd2045895.jpg)

![5768-stripmallzoneevapcooler](https://cloud.githubusercontent.com/assets/5778156/24509697/0cd6fdb4-1535-11e7-9912-07dc05fa3ee1.jpg)

![5768-waterheaterheatpumpwrappedcondenser](https://cloud.githubusercontent.com/assets/5778156/24511181/461476de-1539-11e7-9311-144b8ff8078e.jpg)

rraustad commented:
I admit I do like seeing the mass flow rate changing with inlet conditions for WaterHeaterHeatPumpWrappedCondenser but that's a different issue.

nrel-bot-2 commented:
@rraustad @lgentile it has been 14 days since this pull request was last updated.

nrel-bot-3 commented:
@rraustad @lgentile it has been 14 days since this pull request was last updated.

rraustad committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into #5768-Example-files-which-show-discontinutity-in-air-or-water-flow-rate-through-fans-and-pumps

Conflicts:
	testfiles/StripMallZoneEvapCoolerAutosized.idf

rraustad commented:
Given that #6173 and discussion in #6177 show issues with air loop operation if zone equipment do not have the same inlet/outlet flow rates, this and other changes like this are more important now.

rraustad committed:
Correct indentation

rraustad committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into #5768-Example-files-which-show-discontinutity-in-air-or-water-flow-rate-through-fans-and-pumps

rraustad committed:
Update fan coil unit tests

rraustad commented:
New warning for Fault_FoulingEvapCooler_StripMallZoneEvapCooler (these may required a change to example files to eliminate warnings)

    +   ** Severe  ** InitZoneEvaporativeCoolerUnit: ZoneHVAC:EvaporativeCoolerUnit = SMSTORE6 ZONE EVAP UNIT
    +   **   ~~~   ** ...unit fan volumetric flow rate less than evaporative cooler unit design supply air flow rate.
    +   **   ~~~   ** ...fan volumetric flow rate = 0.50000 m3/s.
    +   **   ~~~   ** ...evap cooler unit volumetric flow rate = 0.65000 m3/s.
    +   **   ~~~   ** ...evaporative cooler unit design supply air flow rate will match fan flow rate and simulation continues.
    +   ** Severe  ** InitZoneEvaporativeCoolerUnit: ZoneHVAC:EvaporativeCoolerUnit = SMSTORE7 ZONE EVAP UNIT
    +   **   ~~~   ** ...unit fan volumetric flow rate less than evaporative cooler unit design supply air flow rate.
    +   **   ~~~   ** ...fan volumetric flow rate = 0.50000 m3/s.
    +   **   ~~~   ** ...evap cooler unit volumetric flow rate = 0.65000 m3/s.
    +   **   ~~~   ** ...evaporative cooler unit design supply air flow rate will match fan flow rate and simulation continues.

For example file HeatPumpWaterHeater, slight reduction in Standard Rated Energy Factor by 0.0004
For example file HeatPumpWaterHeaterStratified,  slight reduction in Standard Rated Energy Factor by 0.0015 ish
For example file StripMallZoneEvapCooler, fan volumetric flow rate increased to match increased user input from 0.5 to 0.65

New warning for StripMallZoneEvapCoolerAutosized

    +   ** Severe  ** InitZoneEvaporativeCoolerUnit: ZoneHVAC:EvaporativeCoolerUnit = SMSTORE6 ZONE EVAP UNIT
    +   **   ~~~   ** ...unit fan volumetric flow rate less than evaporative cooler unit design supply air flow rate.
    +   **   ~~~   ** ...fan volumetric flow rate = 0.50000 m3/s.
    +   **   ~~~   ** ...evap cooler unit volumetric flow rate = 0.65000 m3/s.
    +   **   ~~~   ** ...evaporative cooler unit design supply air flow rate will match fan flow rate and simulation continues.
    +   ** Severe  ** InitZoneEvaporativeCoolerUnit: ZoneHVAC:EvaporativeCoolerUnit = SMSTORE7 ZONE EVAP UNIT
    +   **   ~~~   ** ...unit fan volumetric flow rate less than evaporative cooler unit design supply air flow rate.
    +   **   ~~~   ** ...fan volumetric flow rate = 0.50000 m3/s.
    +   **   ~~~   ** ...evap cooler unit volumetric flow rate = 0.65000 m3/s.
    +   **   ~~~   ** ...evaporative cooler unit design supply air flow rate will match fan flow rate and simulation continues.

No real change to remaining diff's files.

rraustad commented:
Deleted x86_64-Linux-Ubuntu-14.04-gcc-4.8 results to try again.

rraustad commented:
@Myoldmopar the Linux box is showing a fatal for 5ZoneAirCooledWithDOASAirLoop. I don't see any divide by 0 introduced here so maybe it's pointing out some other divide by or something else. Do you feel well enough to run this in debug to see where this is failing?

Myoldmopar commented:
```
Program received signal SIGFPE, Arithmetic exception.
0x00007ffff479d01f in EnergyPlus::SizingManager::UpdateTermUnitFinalZoneSizing ()
    at src/EnergyPlus/SizingManager.cc:3749
3749					thisTUFZSizing.DesCoolOAFlowFrac = min(thisFZSizing.DesCoolOAFlowFrac * minOAFrac / coolFlowRatio, 1.0);
```

Looks like maybe `coolFlowRatio == 0`?

rraustad commented:
You know now I want to be able to run this on my end. Thanks.

rraustad committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into #5768-Example-files-which-show-discontinutity-in-air-or-water-flow-rate-through-fans-and-pumps

rraustad commented:
1) that line is not 3749 in my code, and
2) that line is protected by a 0 check ... so I don't know ...

the first line here is 3749 on my machine, the one you show is 3760:

    thisTUFZSizing.DesHeatVolFlowMax = thisFZSizing.DesHeatVolFlowMax * heatFlowRatio;
    thisTUFZSizing.DesHeatMassFlow = thisFZSizing.DesHeatMassFlow * heatFlowRatio;
    thisTUFZSizing.HeatMassFlow = thisFZSizing.HeatMassFlow * heatFlowRatio;
    thisTUFZSizing.DesHeatMaxAirFlow2 = thisFZSizing.DesHeatMaxAirFlow2 * heatFlowRatio;
    thisTUFZSizing.HeatFlowSeq = thisFZSizing.HeatFlowSeq * heatFlowRatio;
    thisTUFZSizing.HeatLoadSeq = thisFZSizing.HeatLoadSeq * heatLoadRatio;
    thisTUFZSizing.NonAirSysDesHeatLoad = thisFZSizing.NonAirSysDesHeatLoad * heatLoadRatio;
    thisTUFZSizing.NonAirSysDesHeatVolFlow = thisFZSizing.NonAirSysDesHeatVolFlow * heatFlowRatio;
    // Outdoor air
    Real64 minOAFrac = thisTUSizing.SpecMinOAFrac;
    if ( coolFlowRatio > 0.0 ) {
        thisTUFZSizing.DesCoolOAFlowFrac = min( thisFZSizing.DesCoolOAFlowFrac * minOAFrac / coolFlowRatio, 1.0 );
    } else {
        thisTUFZSizing.DesCoolOAFlowFrac = 0.0;
    }

rraustad commented:
SIGFPE is more than divide by right? uninitialized, etc..

Myoldmopar commented:
The last commit in this branch is 4aa7832.  If you look at the code in that commit on Github, on line 3749 of SizingManager [here](https://github.com/NREL/EnergyPlus/blob/4aa7832856d9280e861163ea4dee1d48a22e991f/src/EnergyPlus/SizingManager.cc?utf8=%E2%9C%93#L3749) it shows the suspicious line.  I'm not sure what you are seeing locally, but it is not in sync with this branch.

Myoldmopar commented:
Is there any chance you'd made improvements to the code but not pushed them up?  The last commit on this branch was 15 days ago.

rraustad commented:
Hmmm, I switched to this branch, pulled this branch and then pulled develop. When I diff previous I see the protection of divide by 0. So not sure but will push what I have to see what happens.

rraustad commented:
I pushed, but it looks like only the merge that pushed up.

Myoldmopar commented:
Right, and that's just what it needed.  That fix came in as part of the merge, so this branch is now fully updated with develop (18 commits ahead; 0 behind), so the next round of CI runs will be clean.

rraustad commented:
I assume there is a reason why the Mac sees a 0 node mass flow rate when Windows does not. SecNodeNum = 3.

![piusecmass](https://user-images.githubusercontent.com/5778156/30145580-2c9d810a-9360-11e7-9da6-89a4fe4e7b2e.png)


    EnergyPlusFixture.SeriesPIUTest1 | 1.81246
    Note: Google Test filter = EnergyPlusFixture.SeriesPIUTest1[==========] Running 1 test from 1 test case.[----------] Global test environment set-up.[----------] 1 test from EnergyPlusFixture[ RUN ] 
     EnergyPlusFixture.SeriesPIUTest1/Users/kbenne/ci_runs/EnergyPlus-3679d4651a-x86_64-clang/tst/EnergyPlus/unit/PoweredInductionUnits.unit.cc:459: 
    FailureValue of: DataLoopNode::Node( SecNodeNum ).MassFlowRate  Actual: 0
    Expected: SecMaxMassFlow  Which is: 0.180646

Myoldmopar commented:
I also see a zero locally on Linux.

```
$ ./energyplus_tests --gtest_filter="*SeriesPIUTest1*"
Note: Google Test filter = *SeriesPIUTest1*
[==========] Running 1 test from 1 test case.
[----------] Global test environment set-up.
[----------] 1 test from EnergyPlusFixture
[ RUN      ] EnergyPlusFixture.SeriesPIUTest1
/home/edwin/Projects/energyplus/repos/4eplus/tst/EnergyPlus/unit/PoweredInductionUnits.unit.cc:459: Failure
Value of: DataLoopNode::Node( SecNodeNum ).MassFlowRate
  Actual: 0
Expected: SecMaxMassFlow
Which is: 0.180646
```

rraustad commented:
I need more info. How can I set up a Linux build on a Windows machine?

rraustad commented:
http://metamod-p.sourceforge.net/cross-compiling.on.windows.for.linux.html

Myoldmopar commented:
Easiest way is to get Virtualbox or VMWare installed on your machine, set up a new virtual machine and install Ubuntu 16.04 on it.  I am **happy** to help you out with the process, but it would be Monday before I'll be able to.  Let me know and we can lock in an hour to get you up and running.

rraustad commented:
Thanks, next week or thereafter is fine. I just want to be able to replicate these diff's.

rraustad commented:
This is a little out of my comfort zone. But I think these machines are telling us something that we're missing.

mjwitte commented on src/EnergyPlus/PoweredInductionUnits.cc:
```diff
@@ -1418,6 +1424,7 @@ namespace PoweredInductionUnits {
 			}
 		}
 		if ( GetCurrentScheduleValue( PIU( PIUNum ).SchedPtr ) <= 0.0 ) UnitOn = false;
+		if ( ( GetCurrentScheduleValue( PIU( PIUNum ).FanAvailSchedPtr ) <= 0.0 || DataHVACGlobals::TurnFansOff ) && ! DataHVACGlobals::TurnFansOn ) UnitOn = false;
```
@rraustad Not sure this is correct.  The PIU terminal unit could still pass through primary airflow even if the PIU fan is scheduled off.

rraustad replied:
I'm not positive either but this line of code is in the CalcSeriesPIU so I figured if the fan was off there would always be 0 flow. The TurnFansOn/Off flags are passed to the fan, if these say to turn off the fan, then fan outlet flow rate = 0.

    if ( PIU( PIUNum ).Fan_Num == DataHVACGlobals::FanType_SystemModelObject ) {
        HVACFan::fanObjs[ PIU( PIUNum ).Fan_Index ]->simulate( _, PIUTurnFansOn , PIUTurnFansOff ,_ );
    } else if ( PIU( PIUNum ).Fan_Num == DataHVACGlobals::FanType_SimpleConstVolume ) {
        Fans::SimulateFanComponents( PIU( PIUNum ).FanName, FirstHVACIteration, PIU( PIUNum ).Fan_Index, _, PIUTurnFansOn, PIUTurnFansOff ); // fire the fan
    }
 
To be sure I traced the node connections through the SeriesPIU and see that primary and secondary air flow both pass through the fan. Regardless of real life PIU's, the E+ model does not allow flow through a fan when it is off.

![seriespiuschematic](https://user-images.githubusercontent.com/5778156/30571828-3656ea2c-9cb8-11e7-963f-6e6c7a38bf8c.png)

mjwitte commented on src/EnergyPlus/EvaporativeCoolers.cc:
```diff
@@ -3831,9 +3832,21 @@ namespace EvaporativeCoolers {
+				}
+
 				if ( ZoneEvapUnit( UnitNum ).ActualFanVolFlowRate > 0.0 ) {
 					ZoneEvapUnit( UnitNum ).DesignFanSpeedRatio = ZoneEvapUnit( UnitNum ).DesignAirVolumeFlowRate / ZoneEvapUnit( UnitNum ).ActualFanVolFlowRate;
```
@rraustad Now that the fan is forced to match, DesignFanSpeedRatio can go away.  And the I/O ref should be changed under ZoneHVAC:EvapCooler "Design Supply Air Flow Rate" where it says - This value is typically consistent with the maximum air flow rate of the fan, but it can be less if desired.

rraustad replied:
They are not forced to match, the fan is forced to have a flow greater than or equal to the evap cooler. What if a user hard sizes a fan flow rate greater than the evap cooler? I had just now considered deleting this variable and setting a local FanSpeedRatio = 1 to use in the fan call but then thought of this scenario.

    Fans::SimulateFanComponents( ZoneEvapUnit( UnitNum ).FanName, false, ZoneEvapUnit( UnitNum ).FanIndex, ZoneEvapUnit( UnitNum ).DesignFanSpeedRatio, ZoneCompTurnFansOn, ZoneCompTurnFansOff );

mjwitte replied:
I see, well if they don't match it's still broken for the fan:onoff case. The DesignFanSpeedRatio should *never* be applied to the mass flow rate.  It should only be passed to Fan:OnOff when it's called for the full flow case.  I'll send you the changes I was going to put in before I saw this waiting over here.  You can review and see if you agree.

rraustad replied:
FanSpeedRatio should be passed to the fan based on ratio of mass flow rate to max fan flow rate. So there are multiple cases here for max/min primary flow, sec flow on or off, etc. FanSpeedRatio should be calculated for each case.

mjwitte replied:
But it's only used by Fan:OnOff.  For the variable flow control Fan:OnOff isn't legal.  And for the places where the system is off, sending a zero makes sure the fan is off.

rraustad replied:
Wait a minute, this isn't a multispeed fan, so there is only one speed and therefore only one speed ratio. So yes I would be interested in seeing what changes you have in mind.

rraustad replied:
Yes, if it's a VAV fan then SpeedRatio doesn't matter.

Myoldmopar commented:
@rraustad Email me if you want to try to get a Linux build environment set up tomorrow or Wednesday.  I'm sure we can probably hijack a certain GTM if it isn't in use.  (wink wink @mjwitte).

If you are interested, I'll send you some links to download before we get started, and we'll continue this via email.

rraustad committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into #5768-Example-files-which-show-discontinutity-in-air-or-water-flow-rate-through-fans-and-pumps

rraustad commented:
Given the time crunch, can't we just get someone with a mac to single step that unit test?

rraustad committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into #5768-Example-files-which-show-discontinutity-in-air-or-water-flow-rate-through-fans-and-pumps

rraustad committed:
Address comments and doc update

rraustad committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into #5768-Example-files-which-show-discontinutity-in-air-or-water-flow-rate-through-fans-and-pumps

Myoldmopar committed:
Merge branch 'develop' into #5768-Example-files-which-show-discontinutity-in-air-or-water-flow-rate-through-fans-and-pumps

rraustad committed:
Merge branch '#5768-Example-files-which-show-discontinutity-in-air-or-water-flow-rate-through-fans-and-pumps' of https://github.com/NREL/EnergyPlus into #5768-Example-files-which-show-discontinutity-in-air-or-water-flow-rate-through-fans-and-pumps

rraustad committed:
Fix PIU fan logic and unit test failure

rraustad commented:
PIU unit test was failing because fan control logic wasn't quite right.

rraustad commented:
@Myoldmopar I think the unit test will pass now.

rraustad commented:
@mjwitte I am feeling more comfortable with this now. The PIU unit test uses a fan with an availability schedule AlwaysOff to test the availability managers for air loops (TurnFansOn and TurnFansOff) and zone equipment (TurnZoneFansOnlyOn). Since I had only used TurnFansOn and TurnFansOff in my attempt to correct this the PIU unit test was failing. When I corrected that last night, the existing PIU unit tests pass on the Mac CI now.

Given that the PIU model I corrected is the series model and that the system must be off when the fan is off, as exemplified by the unit test, the line of code in question seems correct to me. This line of code is very similar to what we use for other zone equipment to decide if the fan should operate.

PTHP:

    if ( GetCurrentScheduleValue( PTUnit( PTUnitNum ).SchedPtr ) > 0.0 && ( ( GetCurrentScheduleValue( PTUnit( PTUnitNum ).FanAvailSchedPtr ) > 0.0 || ZoneCompTurnFansOn ) && ! ZoneCompTurnFansOff ) ) {

PIU (this is the corrected version from last night):

    if ( ( GetCurrentScheduleValue( PIU( PIUNum ).FanAvailSchedPtr ) <= 0.0 || PIUTurnFansOff ) && ! PIUTurnFansOn ) UnitOn = false;

where (existing code):

    bool PIUTurnFansOn = ( DataHVACGlobals::TurnFansOn || DataHVACGlobals::TurnZoneFansOnlyOn );  // If True, overrides fan schedule and cycles PIU fan on
    bool PIUTurnFansOff = DataHVACGlobals::TurnFansOff; // If True, overrides fan schedule and PIUTurnFansOn and cycles PIU fan off

So I think this is ready to go now.

Myoldmopar commented:
Yep, unit test passes now, great!  And I am also very comfortable with dropping this in once CI wraps up.
